include/cppkafka/utils/buffered_producer.h

```diff
@@ -188,69 +208,105 @@ class CPPKAFKA_API BufferedProducer {
+    
+    /**
+     * \brief Sets the local message produce failure callback
+     *
+     * This callback will be called when local message production fails during a flush() operation.
+     * Failure errors are typically payload too large, unknown topic or unknown partition.
+     * Note that if the callback returns false, the message will be dropped from the buffer,
+     * otherwise it will be re-enqueued for later retry.
+     *
+     * \param callback
+     *
+     * \warning Do not call any method on the BufferedProducer while inside this callback
+     */
+    void set_flush_failure_callback(FlushFailureCallback callback);
+    
 private:
-    using QueueType = std::queue<Builder>;
+    using QueueType = std::list<Builder>;
+    enum class MessagePriority { Low, High };
 
     template <typename BuilderType>
-    void do_add_message(BuilderType&& builder);
+    void do_add_message(BuilderType&& builder, MessagePriority priority, bool do_flush);
     template <typename MessageType>
     void produce_message(const MessageType& message);
     Configuration prepare_configuration(Configuration config);
     void on_delivery_report(const Message& message);
-
     
+    // Members
     Configuration::DeliveryReportCallback delivery_report_callback_;
     Producer producer_;
     QueueType messages_;
+    mutable std::mutex mutex_;
     ProduceFailureCallback produce_failure_callback_;
-    size_t expected_acks_{0};
-    size_t messages_acked_{0};
+    FlushFailureCallback flush_failure_callback_;
     ssize_t max_buffer_size_{-1};
+    std::atomic_ulong expected_acks_{0};
+    std::atomic_ullong total_messages_acked_{0};
 };
 
 template <typename BufferType>
 BufferedProducer<BufferType>::BufferedProducer(Configuration config)
 : delivery_report_callback_(config.get_delivery_report_callback()),
   producer_(prepare_configuration(std::move(config))) {
-
+    // Allow re-queuing failed messages
+    producer_.set_payload_policy(Producer::PayloadPolicy::PASSTHROUGH_PAYLOAD);
 }
 
 template <typename BufferType>
 void BufferedProducer<BufferType>::add_message(const MessageBuilder& builder) {
-    do_add_message(builder);
+    do_add_message(builder, MessagePriority::Low, true);
 }
 
 template <typename BufferType>
 void BufferedProducer<BufferType>::add_message(Builder builder) {
-    do_add_message(move(builder));
+    do_add_message(move(builder), MessagePriority::Low, true);
 }
 
 template <typename BufferType>
 void BufferedProducer<BufferType>::produce(const MessageBuilder& builder) {
     produce_message(builder);
-    expected_acks_++;
 }
 
 template <typename BufferType>
 void BufferedProducer<BufferType>::produce(const Message& message) {
     produce_message(message);
-    expected_acks_++;
 }
 
 template <typename BufferType>
 void BufferedProducer<BufferType>::flush() {
-    while (!messages_.empty()) {
-        produce_message(messages_.front());
-        messages_.pop();
+    QueueType flush_queue; // flush from temporary queue
```

Comment:
I wonder if there should be a loop here that makes sure that when `flush` exits, any previously added messages are indeed flushed. Right now if you flush and a message fails, it will be left in the queue and only flushed again after `flush` is called again. My expectation is that after calling flush, I actually flushed the writer and everything in it is written.

e.g. 

```c++
while (true) {
    QueueType flush_queue;
    // {  lock & swap }
    if (flush_queue.empty()) {
        return;
    }
    // your loop...
    // wait for acks...
}
```

Reply:
This is the kind of thing i am trying to avoid, if you have this ongoing flush, while other threads (including delivery failures) keep enqueuing at the same or higher rate, you have now created an open faucet situation, where nothing is buffered. My expectation is that 1) flush is called periodically so re-enqueued messages will eventually be delivered 2) flush only flushes what's in the buffer at the time of the call and not behaving like an ongoing flush.

Reply:
Hmm that's fair. I guess if you want to ensure everything is flushed you as a user should loop calling flush while there's still messages pending.